jersey-servers/src/test/java/com/palantir/remoting2/servers/jersey/RequestRateLimiterTest.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting2.servers.jersey;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.Response;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public final class RequestRateLimiterTest {
+
+    private static final long REQUESTS_PER_SECOND = 10;
+    private static final long ALLOWED_DELAY = 1_000 / REQUESTS_PER_SECOND;
+    private static final long NO_DELAY = 0;
+
+    private Function<ContainerRequestContext, Optional<String>> featureFunc;
+    private ArgumentCaptor<Response> responseCaptor = ArgumentCaptor.forClass(Response.class);
+    private ContainerRequestContext request1;
+    private ContainerRequestContext request2;
+    private RequestRateLimiter rateLimiter;
+
+    @Before
+    public void before() throws InterruptedException {
+        request1 = mock(ContainerRequestContext.class);
+        request2 = mock(ContainerRequestContext.class);
+        featureFunc = mock(Function.class);
+        rateLimiter = RequestRateLimiter.create(REQUESTS_PER_SECOND, featureFunc);
+
+        when(featureFunc.apply(request1)).thenReturn(Optional.of("1"));
+        when(featureFunc.apply(request2)).thenReturn(Optional.of("2"));
+    }
+
+    @Test
+    public void testRequestAborted() throws IOException {
+        execute(NO_DELAY, request1, request1);
+
+        verify(request1).abortWith(responseCaptor.capture());
+        assertThat(responseCaptor.getValue().getStatus()).isEqualTo(429);
+    }
+
+    @Test
+    public void testDifferentFeaturesNotLimited() {
+        execute(NO_DELAY, request1, request2);
+
+        verify(request1, never()).abortWith(any());
+        verify(request2, never()).abortWith(any());
+    }
+
+    @Test
+    public void testUnderLimitAllowed() {
+        List<ContainerRequestContext> requests = Collections.nCopies(10, request1);
+        execute(ALLOWED_DELAY, requests);
```

Comment:
think of the limiter as adding tokens to a bucket every 1 / REQUESTS_PER_SECOND, so after the first request I need to wait 1 / PPS until making the next request. The test below that you commented on verifies that if I make a request and then wait 1 full second that I can then make up to the maximum number of requests because I've accrued some number of tokens.